example_notebooks/user_stories/user_story_2/user_story_2_best_case_tre.py

```diff
@@ -0,0 +1,224 @@
+"""
+User story 2 (best case) as TRE
+"""
+
+import argparse
+import logging
+import os
+import pickle
+
+import numpy as np
+import pandas as pd
+from sklearn.preprocessing import LabelEncoder, OneHotEncoder
+
+from aisdc.attacks.attack_report_formatter import GenerateTextReport
+from aisdc.attacks.target import Target  # pylint: disable=import-error
+
+
+def generate_report(
+    directory,
+    target_model,
+    train_indices,
+    test_indices,
+    attack_results,
+    target_filename,
+    outfile,
+):
+    """
+    Generate report based on target model
+    """
+
+    print()
+    print("Acting as TRE...")
+    print(
+        "(when instructions on how to recreate the dataset have been provided by the researcher)"
+    )
+    print()
+
+    filename = directory + target_model
+    print("Reading target model from " + filename)
+    target_model = pickle.load(open(filename, "rb"))
+
+    print("Reading training/testing indices from ./" + directory)
+    indices_train = np.loadtxt(directory + train_indices)
+    indices_test = np.loadtxt(directory + test_indices)
+
+    filename = "user_stories_resources/dataset_26_nursery.csv"
+    print("Reading data from " + filename)
+    data = pd.read_csv(filename)
+
+    print()
+
+    print(data.head())
+    print(indices_test[:10])
+    print(indices_train[:10])
+
+    y = np.asarray(data["class"])
+    x = np.asarray(data.drop(columns=["class"], inplace=False))
+
+    n_features = np.shape(x)[1]
+    indices: list[list[int]] = [
+        [0, 1, 2],  # parents
+        [3, 4, 5, 6, 7],  # has_nurs
+        [8, 9, 10, 11],  # form
+        [12, 13, 14, 15],  # children
+        [16, 17, 18],  # housing
+        [19, 20],  # finance
+        [21, 22, 23],  # social
+        [24, 25, 26],  # health
+    ]
+
+    x_train_orig = np.asarray([x[int(i)] for i in indices_train])
+    y_train_orig = np.asarray([y[int(i)] for i in indices_train])
+    x_test_orig = np.asarray([x[int(i)] for i in indices_test])
+    y_test_orig = np.asarray([y[int(i)] for i in indices_test])
+
+    # Preprocess dataset
+    # one-hot encoding of features and integer encoding of labels
+    label_enc = LabelEncoder()
+    feature_enc = OneHotEncoder()
+    x_train = feature_enc.fit_transform(x_train_orig).toarray()
+    y_train = label_enc.fit_transform(y_train_orig)
+    x_test = feature_enc.transform(x_test_orig).toarray()
+    y_test = label_enc.transform(y_test_orig)
+
+    # Wrap the model and data in a Target object
+    target = Target(model=target_model)
+    target.add_processed_data(x_train, y_train, x_test, y_test)
+    target.add_raw_data(x, y, x_train_orig, y_train_orig, x_test_orig, y_test_orig)
+    for i in range(n_features):
+        target.add_feature(data.columns[i], indices[i], "onehot")
+
+    SAVE_PATH = directory
+
+    # check direct method
```

Comment:
if this is the script for the TRE, why do we need to call the attacks twice (once via the models' run_attack() method and once via request_release()?

Won't that just take extra time?

Reply:
That is a very good point, I was trying to show how to do both just to prove it can be done, but it doesn't add any benefit to this use case. I'll remove the top call - thanks!